clients/shared/merge.go

```diff
@@ -0,0 +1,164 @@
+	"github.com/artie-labs/transfer/lib/optimization"
+	"github.com/artie-labs/transfer/lib/ptr"
+	"github.com/artie-labs/transfer/lib/sql"
+	"github.com/artie-labs/transfer/lib/typing/columns"
+)
+
+func Merge(dwh destination.DataWarehouse, tableData *optimization.TableData, cfg config.Config, opts types.MergeOpts) error {
+	if tableData.ShouldSkipUpdate() {
+		return nil
+	}
+
+	fqName := tableData.ToFqName(dwh.Label(), true, cfg.SharedDestinationConfig.UppercaseEscapedNames, "")
+	tableConfig, err := dwh.GetTableConfig(tableData)
+	if err != nil {
+		return err
+	}
+
+	srcKeysMissing, targetKeysMissing := columns.Diff(tableData.ReadOnlyInMemoryCols(), tableConfig.Columns(),
+		tableData.TopicConfig.SoftDelete, tableData.TopicConfig.IncludeArtieUpdatedAt,
+		tableData.TopicConfig.IncludeDatabaseUpdatedAt, tableData.Mode())
+
+	createAlterTableArgs := ddl.AlterTableArgs{
+		Dwh:               dwh,
+		Tc:                tableConfig,
+		FqTableName:       fqName,
+		CreateTable:       tableConfig.CreateTable(),
+		ColumnOp:          constants.Add,
+		CdcTime:           tableData.LatestCDCTs,
+		UppercaseEscNames: &cfg.SharedDestinationConfig.UppercaseEscapedNames,
+	}
+
+	// Columns that are missing in DWH, but exist in our CDC stream.
+	err = ddl.AlterTable(createAlterTableArgs, targetKeysMissing...)
+	if err != nil {
+		slog.Warn("Failed to apply alter table", slog.Any("err", err))
+		return err
+	}
+
+	// Keys that exist in DWH, but not in our CDC stream.
+	deleteAlterTableArgs := ddl.AlterTableArgs{
+		Dwh:                    dwh,
+		Tc:                     tableConfig,
+		FqTableName:            fqName,
+		CreateTable:            false,
+		ColumnOp:               constants.Delete,
+		ContainOtherOperations: tableData.ContainOtherOperations(),
+		CdcTime:                tableData.LatestCDCTs,
+		UppercaseEscNames:      &cfg.SharedDestinationConfig.UppercaseEscapedNames,
+	}
+
+	if err = ddl.AlterTable(deleteAlterTableArgs, srcKeysMissing...); err != nil {
+		slog.Warn("Failed to apply alter table", slog.Any("err", err))
+		return err
+	}
+
+	tableConfig.AuditColumnsToDelete(srcKeysMissing)
+	tableData.MergeColumnsFromDestination(tableConfig.Columns().GetColumns()...)
+	temporaryTableName := fmt.Sprintf("%s_%s", tableData.ToFqName(dwh.Label(), false, cfg.SharedDestinationConfig.UppercaseEscapedNames, ""), tableData.TempTableSuffix())
+	if err = dwh.PrepareTemporaryTable(tableData, tableConfig, temporaryTableName, types.AdditionalSettings{}); err != nil {
+		return fmt.Errorf("failed to prepare temporary table: %w", err)
+	}
+
+	defer func() {
+		if dropErr := ddl.DropTemporaryTable(dwh, temporaryTableName, false); dropErr != nil {
+			slog.Warn("Failed to drop temporary table", slog.Any("err", dropErr), slog.String("tableName", temporaryTableName))
+		}
+	}()
+
+	// Now iterate over all the in-memory cols and see which one requires backfill.
+	for _, col := range tableData.ReadOnlyInMemoryCols().GetColumns() {
```

Comment:
For BigQuery we used to backfill columns first, then create the temporary table, but now the oder of operations will be reversed.

Reply:
Good flag, our current order should be correct